src/printer.js

```diff
@@ -168,14 +168,23 @@ function genericPrintNoParens(path, options, print) {
     case "LogicalExpression": {
       const parts = [];
       printBinaryishExpressions(path, parts, print, options);
+      const parent = path.getParentNode();
+
+      const shouldNotIndent =
+        (parent.type === "IfStatement" ||
+        parent.type === "WhileStatement" ||
+        parent.type === "DoStatement" ||
+        parent.type === "ForStatement") && n !== parent.body;
+
+      const rest = concat(parts.slice(1));
 
       return group(
         concat([
           // Don't include the initial expression in the indentation
           // level. The first item is guaranteed to be the first
           // left-most expression.
           parts.length > 0 ? parts[0] : "",
-          indent(options.tabWidth, concat(parts.slice(1)))
+          shouldNotIndent ? rest : indent(options.tabWidth, rest)
```

Comment:
You should be able to avoid most of this if you're not indenting, so I'd just have an `if` statement above where you check for if/etc statements and do `return group(concat(parts))` in there.